### PR TITLE
Display Block Information by matching block variations

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description
This PR wants to solve the problem of displaying the proper information of a block, if that block had been created from a `block variation`.

This is achieved with:
1. A new hook (`useMatchingVariationInformation`) which receives a `clientId` and  returns the appropriate block information to display.
2. An addition of a property (`isActive`) in Block Variation's settings. This optional property is a function to match block variations. It doesn't check dynamically all block properties to try to find a match, because in many cases it doesn't make sense. An example is `embed` block where we can have changed the `responsive` attribute, so a match would not be found.

The changes that have been made affect the `Block Card (Inspector Tools)`,  `Navigation List View (top bar)` and `Breadcrumbs`.

To make use of the new hook and handle `Transforms icon`(in toolbar), I need to make `BlockSwitcher` a function component which is handled in this PR: https://github.com/WordPress/gutenberg/pull/27674 to make reviewing easier.
The icon in BlockSwitcher will be handled in a follow up after the refactoring PR lands.

pocVariationsDisplayMatcher

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by ntsekouras** [comment]
This is a comment of @talldan in another draft PR of mine for this problem and I'm just putting it here to have the discussions.

>@ntsekouras This reminds me of a discussion some of us had about the legacy widget block. That block is similar to embeds in that it has variations for the different types of widgets represented. A subset of widgets are 'reference widgets', which can only be used once in a 'post'. The challenge was making the block supports multiple setting work for variations, which was pretty much impossible without being able to reliably identify the variation used to create the block:

This is the issue - #25494

**Comment by ntsekouras** [comment]
@talldan I think if this approach goes on, you could make use of `variationMatcher`.

>With an addition of a property (variationMatcher) in Blocks settings. This optional property is a function to match block variations. It doesn't check dynamically all block properties to try to find a match, because in many cases like it doesn't make sense. An example is embed block where we can have changed the responsive attribute, so a match would not be found.

File: packages/block-editor/src/components/block-card/index.js
Comment (mcsf) [inline packages/block-editor/src/components/block-card/index.js:18]: I don't like that we're making `BlockIcon` and `BlockDescription` so polymorphic. It quickly becomes difficult to understand which level in the element tree is responsible for providing which attributes. At the level of the project, we shouldn't burden developers with choosing between passing `clientId` or explicit values.

**Review by mcsf — COMMENTED** [review]
I like that we're exploring this, but it's a delicate thing to get right. I've left some comments, but I still haven't decided which approach I prefer (e.g. new dedicated selectors, refactor base `getBlockType` selectors, let certain UI components manage it, etc.).

I also haven't commented on the extension of the Variations API with `variationMatcher`, but I'd like us to weigh it against different options (e.g. `blockType.variationAttributeName`, `blockType.variations[0].isActive()`, etc.).

File: packages/block-editor/src/components/block-card/index.js
Comment (gziolo) [reply packages/block-editor/src/components/block-card/index.js:18]: The initial idea I had was to unify the way how those leaf components are used. In particular, I wanted to make it easier to use in a different context where you could apply some more advanced modifications inside the component:

```js
return (
	<div className="block-editor-block-card">
		<BlockIcon clientId={ clientId } showColors />
		<div className="block-editor-block-card__content">
			<BlockTitle clientId={ clientId } as="h2" className="block-editor-block-card__title">
			<BlockDescription clientId={ clientId } className="block-editor-block-card__description" />
		</div>
	</div>
);
```

In practice, the same `BlockCard` component is used not only in the block sidebar but also in the block preview in the inserter where you don't have access to `clientId` yet.

I personally feel like developers shouldn't be concerned about deciding how for a given block to get the icon (applies to the title and description as well) and pass it to the matching component. It should accept either `clientId` or `name` and do the rest. If you think that the opposite makes more sense also because of the backward compatibility then we will have to find a way to ensure that correct data is being passed in all places. The biggest challenge is the ambiguity we introduce by having `getBlockType` that returns the original block definition and the new method that combines that with the active block variation that can be only achieved by providing the `clientId`.

File: packages/block-editor/src/components/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/index.js:94]: I think if we follow this or with some changes direction we should expose it. It should look like @gziolo example:
`<BlockTitle clientId={ clientId } as="h2" className="block-ed....." />` to handle the Title, Description and Icon in a consistent way.

File: packages/blocks/src/store/selectors.js
Comment (ntsekouras) [reply packages/blocks/src/store/selectors.js:131]: You're absolutely right about that.

File: packages/blocks/src/store/selectors.js
Comment (ntsekouras) [reply packages/blocks/src/store/selectors.js:110]: I think so too, but mostly because of the `label`. It probably should be better to handle label where needed, but other than that I don't believe using two stores are a lot..

Before combining them I had a selector in `block` store similar to `getBlockType` but also checking for variations and each of the components (BlockIcon etc..) would retrieve the other info (attributes) from the `block-editor` store. The exact same code..

File: packages/block-editor/src/components/block-card/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-card/index.js:18]: >At the level of the project, we shouldn't burden developers with choosing between passing clientId or explicit values.

I personally don't think it's will be a burden to devs, quite the opposite.. This way we have a consistent way(using the three components) of displaying block's info and could even be more easy to use with the fallback of getting the `selectedBlockClientId`. For example if we need to show info on a selected block it could be just `<BlockDescription />`.

**Comment by ntsekouras** [comment]
Thanks for starting to look at it and sharing your feedback @gziolo and @mcsf - really appreciate it. This is a problem that we would need to solve (matching variation) and by discussing it, will find the best approach!

I know that the current status of the PR (as of writing this comment) is not the best, but it's a start :) I'd like some feedback from others as well that have worked in related things. --cc @talldan, @youknowriad , @noahtallen

File: packages/block-editor/src/components/block-card/index.js
Comment (mcsf) [reply packages/block-editor/src/components/block-card/index.js:18]: I think you two make good arguments, leaving me undecided. I have a follow-up question, but let me know if it's too farfetched:

What if `BlockIcon`, BlockTitle`, `BlockDescription` are made to be contextual:

```jsx
<BlockEditContextProvider value={ ... }> /* implicit ancestor */
    <BlockIcon showColors />
    <BlockTitle as="h2" />
```

Comment (mcsf) [reply packages/block-editor/src/components/block-card/index.js:18]: (edit: I accidentally sent this comment before finishing it)

I think you two make good arguments, leaving me undecided. I have a follow-up question, but let me know if it's too farfetched:

What if `BlockIcon`, `BlockTitle`, `BlockDescription` are made to be contextual:

```jsx
<BlockEditContextProvider value={ ... }> /* implicit ancestor */

        <div className="block-editor-block-card">
            <BlockIcon showColors />
            <BlockTitle as="h2" />
```

... with the provision that `BlockIcon` accepts an explicit `blockName` prop as an override:

```js
BlockIcon = ( { blockName } ) => {
    const { clientId } = useBlockEditContext();
    const icon = blockName ?
        grabIconFromBlockType( blockName ) :
        grabIconFromBlockInstance( clientId );
    /* ... */
```

That way, `BlockCard` might choose to provide its own context in order to address:

> In practice, the same BlockCard component is used not only in the block sidebar but also in the block preview in the inserter where you don't have access to clientId yet.

```js
BlockCard = ( props ) => {
    const element = (
        <div className="block-editor-block-card">
            <BlockIcon showColors />
            <BlockTitle as="h2" />
        </div>
    );

    if ( empty( props ) ) {
        return element;
    }

    return (
        <BlockEditContextProvider value={ props }>
             { element }
        </BlockEditContextProvider>       
    );
}
```

Comment (gziolo) [reply packages/block-editor/src/components/block-card/index.js:18]: I wouldn't mind if we would take advantage of `useBlockEditContext` here, we do it in other places in `@wordpress/block-editor` since it holds both the block's `name` and `clientId`.

https://github.com/WordPress/gutenberg/blob/18349c49bab0d9a227e2608a66420f7a9f6e70ac/packages/block-editor/src/components/block-edit/context.js#L12-L16

Comment (ntsekouras) [reply packages/block-editor/src/components/block-card/index.js:18]: Me neither! I can try these changes, along with some of the other comments and take it from there.. :)

Comment (ntsekouras) [reply packages/block-editor/src/components/block-card/index.js:18]: Me neither! I can try these changes, along with some of the other comments and take it from there.. :)

**Comment by gziolo** [comment]
> I also haven't commented on the extension of the Variations API with `variationMatcher`, but I'd like us to weigh it against different options (e.g. `blockType.variationAttributeName`, `blockType.variations[0].isActive()`, etc.).

I personally lean towards `blockType.variations[0].isActive()`. This API is specific to variations so it might be difficult to explain in documentation when added as a top-level block type property. While it might create some code duplication when added to individual variations but it's perfectly fine to use a shared function. Besides, it gives the opportunity to select which variations should be considered for the check and even it can be customized per individual item.

I was also wondering, whether the `getBlockDisplayInformation` selector still necessary now that there is a hook that consolidates the same functionality? @youknowriad, can you share some recommendations in terms of performance as it needs to use data from two stores: `core/blocks` and `core/block-editor`. It would be also recomputed every time a block attribute changes, a block type or its block variations.

**Comment by ntsekouras** [comment]
I have created a separate PR to refactor `BlockSwitcher` there, to make this one easier to review: https://github.com/WordPress/gutenberg/pull/27674.

>I personally lean towards blockType.variations[0].isActive(). This API is specific to variations so it might be difficult to explain in documentation when added as a top-level block type property. While it might create some code duplication when added to individual variations but it's perfectly fine to use a shared function. Besides, it gives the opportunity to select which variations should be considered for the check and even it can be customized per individual item.

I believe that the main use case of matching block variations doesn't need much customization per variation. I can't think of an example that could utilize that, but of course there could be...

What seems to me important though is:
>This API is specific to variations so it might be difficult to explain in documentation when added as a top-level block type property.

So I'm going to make it per variation.

**Comment by ntsekouras** [comment]
>I was also wondering, whether the getBlockDisplayInformation selector still necessary now that there is a hook that consolidates the same functionality?

I removed the selector. I'll need to revisit `BlockTitle` though because we have some logic there that needs the `blockType` and `attributes` for `getBlockLabel`. I haven't decided how to handle this yet...

File: packages/block-editor/src/components/block-title/index.js
Comment (gziolo) [inline packages/block-editor/src/components/block-title/index.js:61]: Should it use the block information here as well?

```suggestion
		return `${ blockInformation.title }: ${ truncate( label, { length: 15 } ) }`;
```

Comment (ntsekouras) [reply packages/block-editor/src/components/block-title/index.js:61]: I'm not really sure.. I want to revisit all this `label` functionality, but I want to do it in a follow up to investigate properly and not include possibly even more changes in this PR.

For now by leaving it as is `blockType.title` it keeps the exact previous functionality for cases that `label` was used.

Comment (gziolo) [reply packages/block-editor/src/components/block-title/index.js:61]: Sure, I just wanted to flag it.

File: packages/block-editor/src/components/use-matching-variation-information/index.js
Comment (gziolo) [inline packages/block-editor/src/components/use-matching-variation-information/index.js:40]: Both `getBlockType` and `getBlockVariations` are only tiny wrappers over `core/blocks` datastore calls targeted mostly for plugin authors. It would be more convenient to use direct store references inside `useSelect` calls. It should allow running all the logic inside `useSelect` call.

File: packages/blocks/src/api/registration.js
Comment (gziolo) [inline packages/blocks/src/api/registration.js:142]: Block types acceppt `WPIcon` that is quite complex structure:
```suggestion
 * @property {WPIcon|JSX.Element} icon        Block type icon.
```

File: packages/blocks/src/api/registration.js
Comment (gziolo) [inline packages/blocks/src/api/registration.js:137]: It feels more like part of `@wordpress/block-editor` package. Are you sure it should be promoted as a subset of `WPBlock` type? Isn't it a type that exists only for UI purposes that is completely different?

File: packages/blocks/src/api/registration.js
Comment (gziolo) [inline packages/blocks/src/api/registration.js:100]: There is some issue with indentation when browsing on GitHub.

File: packages/block-library/src/embed/variations.js
Comment (gziolo) [inline packages/block-library/src/embed/variations.js:349]: It would be more convenient when `variationAttributes` would be passed as the second param so people don't have to repeat hard coded values when they define this function in the variation object.

```js
isActive( blockAttributes, variationAttributes ) {
     return blockAttributes.providerNameSlug === variationAttributes.providerNameSlug;
},
```

**Review by gziolo — COMMENTED** [review]
Nice work, I think it's close 👍🏻 

There are some things to clarify:
- what should be the final API for `isActive` matcher added to the block variation
- how should `useMatchingVariationInformation` be called, the type proposed is `WPBlockDisplayInformation` so it needs some alignment, it feels like `useBlockDisplayInformation` would better express intent
- we should measure the performance implications of `useMatchingVariationInformation`, at the moment it probably will recompute every time `core/block-editor` changes which would be good, we should ensure it doesn't re-render when other stores change.

File: packages/blocks/src/api/registration.js
Comment (ntsekouras) [reply packages/blocks/src/api/registration.js:138]: You're probably right. I'm thinking of naming it differently, move it to `block-editor` and have it return the `attributes` and `blockType` as well. That would simplify the usage of `getBlockLabel` at least in `BlockTitle` and maybe in more places when we'll check this `label` functionality`.

What do you think of having it return these extra props (attributes,blockType)?

Comment (gziolo) [reply packages/blocks/src/api/registration.js:138]: I don't know much about `getBlockLabel`, can we leave it out for now?

Comment (ntsekouras) [reply packages/blocks/src/api/registration.js:138]: Yeah, sure!

**Comment by ntsekouras** [comment]
>what should be the final API for isActive matcher added to the block variation

Changed like this per suggestion: https://github.com/WordPress/gutenberg/pull/27469#discussion_r544145212

>how should useMatchingVariationInformation be called, the type proposed is WPBlockDisplayInformation so it needs some alignment, it feels like useBlockDisplayInformation would better express intent

Makes sense and changed it to `useBlockDisplayInformation`

File: packages/block-editor/src/components/use-block-display-information/index.js
Comment (gziolo) [inline packages/block-editor/src/components/use-block-display-information/index.js:37]: Is it possible to move all logic to `useSelect`?

```js
export default function useBlockDisplayInformation( clientId ) {
    return useSelect(
        // all logic here
    );
}
```

Comment (ntsekouras) [reply packages/block-editor/src/components/use-block-display-information/index.js]: moved here: https://github.com/WordPress/gutenberg/pull/27469/commits/a42d2b22288484aebf42b6dea8ac30656bbab7ed

File: packages/block-editor/src/components/index.js
Comment (gziolo) [inline packages/block-editor/src/components/index.js:121]: Any reason why it should be experimental? I personally can't think of anything missing so it cab be considered final.

File: packages/block-editor/src/components/use-block-display-information/README.md
Comment (gziolo) [inline packages/block-editor/src/components/use-block-display-information/README.md:3]: Typo:
```suggestion
A React Hook that tries to find a matching block variation and returns the appropriate information for display reasons. In order to try to find a match we need two things:
```

File: packages/block-editor/src/components/use-block-display-information/README.md
Comment (gziolo) [inline packages/block-editor/src/components/use-block-display-information/README.md:5]: Typo:
```suggestion
1. Block's client id to extract its current attributes.
```

File: packages/block-editor/src/components/use-block-display-information/README.md
Comment (gziolo) [inline packages/block-editor/src/components/use-block-display-information/README.md:6]: How do you like the following version?
```suggestion
2. A block variation has `isActive` prop defined with a matcher function.
```

File: packages/block-editor/src/components/use-block-display-information/README.md
Comment (gziolo) [inline packages/block-editor/src/components/use-block-display-information/README.md:14]: A proposed revision to this sentence:
```suggestion
The hook returns an object which contains the block's title, icon, and description. If no block type is found for the provided `clientId`, it returns `null`.
```

File: packages/block-editor/src/components/use-block-display-information/README.md
Comment (gziolo) [inline packages/block-editor/src/components/use-block-display-information/README.md:26]: Is it some copy/paste issue?
```suggestion
		<div>
```

File: packages/block-editor/src/components/use-block-display-information/index.js
Comment (gziolo) [inline packages/block-editor/src/components/use-block-display-information/index.js:32]: Nit, this line is not needed.
```suggestion
```

File: packages/block-editor/src/components/use-block-display-information/index.js
Comment (gziolo) [inline packages/block-editor/src/components/use-block-display-information/index.js:31]: Should it be more specific? Example:
```suggestion
 * @return {?WPBlockDisplayInformation} Block's display information, or `null` when the block or its type not found.
```

File: packages/block-editor/src/components/use-block-display-information/index.js
Comment (gziolo) [inline packages/block-editor/src/components/use-block-display-information/index.js:40]: Can we use the store object rather than the hardcoded string? It's planned for the existing code in #27088 anyway.

File: packages/block-editor/src/components/use-block-display-information/index.js
Comment (gziolo) [inline packages/block-editor/src/components/use-block-display-information/index.js:52]: Using store action you can remove variations so this check should account for an empty array as well. Technically, it would still work but there is some further processing that can be avoided.

**Review by gziolo — COMMENTED** [review]
This PR is in excellent shape now so I left mostly nitpicks. Thank you for all iterations. I'm personally satisfied with the last revision proposed.  I still need to test this functionality although I anticipate it being only a formal step because e2e tests added cover that 👍🏻 

I will wait for your response before doing the final check and approve this PR.

File: packages/block-editor/src/components/use-block-display-information/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/use-block-display-information/index.js:51]: This is a good catch! 👍

File: packages/block-editor/src/components/index.js
Comment (gziolo) [reply packages/block-editor/src/components/index.js]: Now that happened, [CHANGELOG](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/CHANGELOG.md) could benefit from a note as well. It doesn't feel like it's updated frequently :(

**Comment by ntsekouras** [comment]
Thanks for all the help here @gziolo! I addressed all the feedback and it's waiting for the final check to land 🎉  😄

**Comment by gziolo** [comment]
There is one issue I discovered, the icon in the block toolbar is coming from the block type rather than from the variation:

Screen Shot 2020-12-23 at 13 58 58

Screen Shot 2020-12-23 at 13 58 58

In all other places, I'm aware of, everything works perfectly fine. 

I did some testing with the profiler I couldn't notice any unnecessary re-renders in the block card when I was changing unrelated block attributes.

Stats from the performance e2e tests look good as well:

Screen Shot 2020-12-23 at 13 55 09

Screen Shot 2020-12-23 at 13 55 14

**Comment by ntsekouras** [comment]
>There is one issue I discovered, the icon and its label in the block toolbar is coming from the block type rather than from the variation:

Thanks for testing performance impact for this! 💯 

@gziolo I have it mind and that's the reason. I will create a follow up as soon as this one lands. Sounds okay with you?


From my PR description:
>To make use of the new hook and handle Transforms icon(in toolbar), I need to make BlockSwitcher a function component which is handled in this PR: #27674 to make reviewing easier.
The icon in BlockSwitcher will be handled in a follow up after the refactoring PR lands.

**Review by gziolo — APPROVED** [review]
🚢

**Pull request merged**
